### PR TITLE
Setup the project

### PR DESCRIPTION
**Pull request by andyduong1920**
https://github.com/anduonghien/ex_marketer/projects/1#card-41739022

## What happened

Setup the project

## Insight

1/ Setup the Phoneix project in Umbrella mode

image

2/ Setup credo

3/ Setup sobelow

4/ Setup code coverage

5/ Setup Github Action

6/ Setup Heroku Application

## Proof Of Work

1/ CI Passed

image

2/ CI Passed

image

3/ Heroku app https://ex-marketer.herokuapp.com/

### CODE REVIEW

File: .github/workflows/ci-workflow.yml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    - uses: actions/checkout@v2
+      with:
+        ref: ${{ github.head_ref }}
+
+    - name: Update container
+      run: |
+        apk update
+        apk add build-base
+        mix local.rebar --force
+        mix local.hex --force
```
Comment (olivierobert) [inline .github/workflows/ci-workflow.yml:31]: Line break needed here

File: .github/workflows/ci-workflow.yml [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    - name: Install Dependencies
+      run: mix deps.get
+
+    - name: Create database
+      run: mix ecto.create
+      env:
+        MIX_ENV: test
+        DB_HOST: db
+
+    - name: Migrate database
```
Comment (olivierobert) [inline .github/workflows/ci-workflow.yml:42]: FYI we updated the Github action workflow on Omise Labs to use https://github.com/actions/cache to cache the build app to speed up tests. We will include it in the Elixir template during the Growth project.

File: config/prod.secret.exs [hunk 3]
```diff
@@ -0,0 +1,41 @@
+# In this file, we load production configuration and secrets
```
Comment (olivierobert) [inline config/prod.secret.exs:1]: Normally this file should not be committed to the git repo. This must be removed from the template too 🙈

**Review by olivierobert — COMMENTED** [review]
Looks good. Not sure why you went the umbrella apps way as the test application is small in terms of features. Do you plan to have one main web app and several other OTP apps?

File: .github/workflows/ci-workflow.yml (hunk 2 above)
Comment (andyduong1920) [reply .github/workflows/ci-workflow.yml:42]: Wow, let see how it works then, never work with Github action much, once that merged to Elixir template, I will adjust on this PR

File: config/prod.secret.exs (hunk 3 above)
Comment (andyduong1920) [reply config/prod.secret.exs:1]: Actually, in my understanding, it's not correct @olivierobert If we hard-code the value in this file then we need to add it to `.gitignore`

As here, I'm using it to load the `env` variable (following base Phonenix setup) also as my app is deploy to Heroku (don't Dockerize the app) so I need this file in order to build the app on Heroku @olivierobert 


https://hexdocs.pm/phoenix/deployment.html#handling-of-your-application-secrets

Pasted_Image_7_14_20__17_22

**Comment by andyduong1920** [comment]
> Looks good. Not sure why you went the umbrella apps way as the test application is small in terms of features. Do you plan to have one main web app and several other OTP apps?

Yes @olivierobert I prefer to made this as an `umbrella apps` as it's easy to scale later, I have some idea to build this stuff like can build a `CLI app` or another module later.

Also as my understanding, the Elixir community prefer to separate the Web related layer between the core business so most of the app following the Umbrella mode, it makes the code isolate, something relative to business logic will be in the core app, and the web app just contains the web stuff.

File: config/prod.secret.exs (hunk 3 above)
Comment (olivierobert) [reply config/prod.secret.exs:1]: Heroku does support container deployments :-) Not needed for this specific project but using env variables + [Phoenix releases](https://hexdocs.pm/phoenix/releases.html) would be the best to go.

Comment (andyduong1920) [reply config/prod.secret.exs:1]: Well if we design to go with https://hexdocs.pm/phoenix/releases.html then we could skip this file ;) 

I think we should add it to our `elixir-template` or Comppas that all Elixir projects should go with Release process, what do you think? so we won't have the conflict on each project 🌻

**Comment by olivierobert** [comment]
@anduonghien Regarding umbrella apps, actually after working on such app for a few months, I can tell you that it works entirely different than Rails engines 😆 Also umbrella apps do come with more complex tooling.

A regular Phoneix app already separates the core business logic layer (`lib/app`) from the web layer (`lib/_app_web`). Also, Phoenix contexts allow splitting by domain quite well IMO. There is no need for umbrella apps for that. 

The purpose of umbrella apps is mainly to deploy each app independently. Even in local development, each app should run on a different port. Technically, it's also possible with our engine structures to deploy each engine separately but the main app is shared. Note that it's possible to mix a Phoenix app with other OTP app (e.g. CLI app or else) and in that case it makes sense. But having multiple Phoenix apps in `apps/` is usually with the purpose to deploy each independently.

File: config/prod.secret.exs (hunk 3 above)
Comment (olivierobert) [reply config/prod.secret.exs:1]: That's the goal with our Growth project 👉  https://trello.com/c/GN3EPBtO 😆

**Review by olivierobert — APPROVED** [review]

**Pull request merged**